api/reaction.go

```diff
@@ -0,0 +1,199 @@
+	}
+
+	channelId := mux.Vars(r)["channel_id"]
+	if len(channelId) != 26 {
+		c.SetInvalidParam("saveReaction", "postId")
+		return
+	}
+
+	if !HasPermissionToChannelContext(c, channelId, model.PERMISSION_READ_CHANNEL) {
+		return
+	}
+
+	postId := mux.Vars(r)["post_id"]
+	if len(postId) != 26 || postId != reaction.PostId {
+		c.SetInvalidParam("saveReaction", "postId")
+		return
+	}
+
+	pchan := Srv.Store.Post().Get(reaction.PostId)
+
+	var postHadReactions bool
+	if result := <-pchan; result.Err != nil {
+		c.Err = result.Err
+		return
+	} else if post := result.Data.(*model.PostList).Posts[postId]; post.ChannelId != channelId {
+		c.Err = model.NewLocAppError("saveReaction", "api.reaction.save_reaction.mismatched_channel_id.app_error",
+			nil, "channelId="+channelId+", post.ChannelId="+post.ChannelId+", postId="+postId)
+		c.Err.StatusCode = http.StatusBadRequest
+		return
+	} else {
+		postHadReactions = post.HasReactions
+	}
+
+	if result := <-Srv.Store.Reaction().Save(reaction); result.Err != nil {
+		c.Err = result.Err
+		return
+	} else {
+		go sendReactionEvent(model.WEBSOCKET_EVENT_REACTION_ADDED, channelId, reaction, postHadReactions)
+
+		reaction := result.Data.(*model.Reaction)
+
+		w.Write([]byte(reaction.ToJson()))
+	}
+}
+
+func deleteReaction(c *Context, w http.ResponseWriter, r *http.Request) {
+	reaction := model.ReactionFromJson(r.Body)
+	if reaction == nil {
+		c.SetInvalidParam("deleteReaction", "reaction")
+		return
+	}
+
+	if reaction.UserId != c.Session.UserId {
+		c.Err = model.NewLocAppError("deleteReaction", "api.reaction.delete_reaction.user_id.app_error", nil, "")
+		c.Err.StatusCode = http.StatusForbidden
+		return
+	}
+
+	channelId := mux.Vars(r)["channel_id"]
+	if len(channelId) != 26 {
+		c.SetInvalidParam("deleteReaction", "postId")
+		return
+	}
+
+	if !HasPermissionToChannelContext(c, channelId, model.PERMISSION_READ_CHANNEL) {
+		return
+	}
+
+	postId := mux.Vars(r)["post_id"]
+	if len(postId) != 26 || postId != reaction.PostId {
+		c.SetInvalidParam("deleteReaction", "postId")
+		return
+	}
+
+	pchan := Srv.Store.Post().Get(reaction.PostId)
+
+	var postHadReactions bool
+	if result := <-pchan; result.Err != nil {
+		c.Err = result.Err
+		return
+	} else if post := result.Data.(*model.PostList).Posts[postId]; post.ChannelId != channelId {
+		c.Err = model.NewLocAppError("deleteReaction", "api.reaction.delete_reaction.mismatched_channel_id.app_error",
+			nil, "channelId="+channelId+", post.ChannelId="+post.ChannelId+", postId="+postId)
+		c.Err.StatusCode = http.StatusBadRequest
+		return
+	} else {
+		postHadReactions = post.HasReactions
+	}
+
+	if result := <-Srv.Store.Reaction().Delete(reaction); result.Err != nil {
+		c.Err = result.Err
+		return
+	} else {
+		go sendReactionEvent(model.WEBSOCKET_EVENT_REACTION_REMOVED, channelId, reaction, postHadReactions)
+
+		ReturnStatusOK(w)
+	}
+}
+
+func sendReactionEvent(event string, channelId string, reaction *model.Reaction, postHadReactions bool) {
```

Comment:
We have the post, but it's possibly outdated since `post.HasReactions` may have changed when the reaction was added/removed, so we still need to get it from the DB to check that. I could change the store call to return both the saved reaction and the updated post, but that would make it work differently than every other DB call in the app.